core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala

```diff
@@ -0,0 +1,146 @@
+ *
+ * In this case, provided taskId will be ignored. The benefit for taskId unrelated strategy is that
+ * different taskSets can learn experience from other taskSet to avoid allocating tasks on
+ * problematic executors.
+ */
+private[scheduler] class SimpleStrategy(
+    maxFailureTaskNumber: Int,
+    maxBlackExecutorNumber: Int,
+    val expireTimeInMilliseconds: Long
+  )extends BlacklistStrategy {
+
+  private def getSelectedExecutorMap(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus]) = {
+    executorIdToFailureStatus.filter{
+      case (id, failureStatus) => failureStatus.totalNumFailures > maxFailureTaskNumber
+    }
+  }
+
+  // As this is a taskId unrelated strategy, the input taskId will be ignored
+  def getExecutorBlacklist(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus],
+      atomTask: BlacklistAtomTask, clock: Clock): Set[String] = {
+    getSelectedExecutorMap(executorIdToFailureStatus).keys.toSet
+  }
+
+  def getNodeBlacklist(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus]): Set[String] = {
+    getSelectedExecutorMap(executorIdToFailureStatus)
+      .groupBy{case (id, failureStatus) => failureStatus.host}
+      .filter {case (host, executorIdToFailureStatus) =>
+        executorIdToFailureStatus.size > maxBlackExecutorNumber}
+      .keys.toSet
+  }
+}
+
+/**
+ * This strategy is applied as default to keep the same semantics as original. It's an taskId
+ * related strategy. If an executor failed running "task A", then we think this executor is
+ * blacked for "task A". And we think the executor is still healthy for other task. node blacklist
+ * is always empty.
+ *
+ * It was the standard behavior before spark 1.6
+ */
+private[scheduler] class DefaultStrategy(
+    val expireTimeInMilliseconds: Long) extends BlacklistStrategy {
+  def getExecutorBlacklist(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus],
+      atomTask: BlacklistAtomTask, clock: Clock): Set[String] = {
+    executorIdToFailureStatus.filter{
+      case (_, failureStatus) => failureStatus.numFailuresPerTask.keySet.contains(atomTask) &&
+        clock.getTimeMillis() - failureStatus.updatedTime < expireTimeInMilliseconds
+    }.keys.toSet
+  }
+
+  def getNodeBlacklist(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus]): Set[String] =
+        Set.empty[String]
+}
+
+/**
+ * Create BlacklistStrategy instance according to SparkConf
+ */
+private[scheduler] object BlacklistStrategy {
+  def apply(sparkConf: SparkConf): BlacklistStrategy = {
+    val timeout = sparkConf.getTimeAsMs("spark.scheduler.blacklist.timeout",
+        sparkConf.getLong("spark.scheduler.executorTaskBlacklistTime", 0L).toString() + "ms")
+    sparkConf.get("spark.scheduler.blacklist.strategy", "default") match {
+      case "default" =>
+        new DefaultStrategy(timeout)
+      case "threshold" =>
+        new SimpleStrategy(
```

Comment:
can you rename 
"default" -> "SingleTask"
"threshold" -> "ExecutorAndNode"
, both the conf names and the class names?
I don't love those names either, but I think "default" and "threshold" are not clear at all, so am just trying to choose something better.